### ISSUE TITLE
Support mobile phone based two-factor authentication

### ISSUE THREAD

**Issue by oberstet** [tavendo/WAMP#143]
Specifically, the goal is to allow authenticating WAMP connections using a username, a password ("something that the user knows") and a smartphone with a OTP app ("something that the user possesses"), such as [Google Authenticator](http://en.wikipedia.org/wiki/Google_Authenticator).

Ideally, this should combine:

* WAMP-SCRAM
* WAMP-TOTP

**Comment by Paranaix** [comment]
This is only slightly related to this specific issue but maybe we should change the policy for authmethods. There should be a core set of mandatory authentication methods every router must implement for protocol confirmness. IMO such core set should include *WAMP-CRA*, *WAMP-SCRAM*, *ticket* and *TLS*.  Which other authentication methods a router accepts besides those should be up to him. This might e.g also include fully custom authmethods which are defined by a user of the router through a specific interface.

Now to the actual topic: IMO two-factor auth shouldn't be included in yonder core set but should be router specific. This is due to the fact, that two-factor auth needs specific knowledge about the actual implementation. Usually communication to specific 3rd party services is required. One router might support App A, another one might support App B and another one maybe sends some SMS code, who knows for sure. But all would call that authentication method "WAMP-two-factor", though they are actually completely different and not interchangeable.

Instead we should give the underlying (WAMP) messages required for authentication enough freedom to implement a wide variety of possible authentication methods, including two-factor and WAMP-SCRAM. Specific two-factor auth implementations, e.g one which which uses Google Authenticator should then be implemented by the router but should not be defined by the standard.

**Comment by oberstet** [comment]
I agree, a "generic" 2-factor authentication *method* does not make sense, as "2 factor auth" is just a description of a whole class of auth methods. And there are just too many specific ones and combinations in particular.

However, I think it would be good if we would describe a "best practice" mechanism that people can rely upon, instead of "only" providing a kind of building kit from which developers then have to chose and combine.

In other words, so we can provide an answer to "What authentication scheme *should* I ideally use?"

E.g. most of the time, I'd like to authenticate clients using `["cookie", "scram-totp"]`.

That is, first try cookie based authentication. This covers browsers where the user already had authenticated previously. The cookie can be expired according to different policies. 

If no cookie is set (or the cookie expired), do SCRAM-TOTP. That is a 2-factor authentication that combines SCRAM for password-login, and requiring TOTP in addition.

**Comment by oberstet** [comment]
From Yury on the maling list:

> ... In the meantime, I have been
> experimenting with how to add OTP functionality in the minimally
> invasive way. Here are my ideas so far:
> 
> 1. By agreement, both the server and the client use
> pbkdf2(original_hash, otp) as the new password hash. This is a
> horrible hack, however it requires no changes to the current
> infrastructure.
> 
> 2. Piggyback the OTP inside the WAMP-CRA authentication using the
> 'extra' field. This is not very elegant, but would be the easiest to
> implement router side. It would require the client to return something
> like types.Signature(signature, otp) from onChallange. This would be
> serialized to message.Authenticate(signature, otp), sent across the
> wire, and made available to the authenticator. This would be backwards
> compatible with the current onChallenge API provided the return type
> is checked.
> 
> 3. Allow several authentication-challenge steps. I think this is the
> ideal solution, but it is more difficult to implement. When the router
> receives HELLO, it sends out several challenges. It would send out a
> WAMP-CRA challenge, and an OTP challenge, for example. These can be
> distinguished by having an 'authmethod' field set inside 'extra'. The
> client would respond to both challenges preserving the 'authmethod'
> part of 'extra'. The router would then collect the responses, decide
> on an authrole and authid, and welcome or deny the client. You can
> even allow elaborate chains (sort of like PAM) where one
> authentication method may be sufficient (cookie or ticket), or another
> may be required (IP filter).

**Comment by ysobolev** [comment]
I am partial to 3. It is significantly more general and seems less like a hack. Both 2 and 3 require the same change client side. Namely, we need the ability to specify the 'extra' parameter when responding inside onChallenge.

On the router side, things get tricky. Number 3 would require keeping track of the client state in more detail. For crossbar, I propose sending the result from every onAuthenticate to a "super-authenticator." This would be a custom method that either returns a types.Accept, types.Deny, or a (new) types.AuthenticationPending. This can do bookkeeping to see if every authentication method has fired. A default one can be supplied. If a user so desires, they can implement a PAM-like system here.

**Comment by ysobolev** [comment]
After some discussion on IRC, some problems become apparent. First of all, WAMP-SCRAM has four sequential parts: client-first, server-first, client-last, server-last. The client authenticates to the server in client-last, and the server authenticates to the client in server-last. In a pure WAMP-SCRAM setting, this can map to WAMP messages as follows:

HELLO (client-first)
CHALLENGE (server-first)
AUTHENTICATE (client-last)
WELCOME (server-last)

Unfortunately, this is not possible if there is more than one authentication method. The client may not wish to send an OTP or further authentication until is has received the server-last confirming the server's identity.

This problem can be remedied by adding another WAMP message type. Obviously, it would be best to avoid this.

Additionally, there are at least two ways to implement multi-factor authentication: sequentially and in parallel. These are detailed below.

1. Send challenges out sequentially.
For example, the authentication flow would look like:
C: HELLO [scram-client-first]
S: CHALLENGE [scram-server-first]
C: AUTHENTICATE [scram-client-last]
S: COUNTER-AUTHENTICATE [scram-server-last] (This is a necessary step in SCRAM where the server proof is sent. It can potentially be bundled with the following challenge.)
S: CHALLENGE [totp]
C: AUTHENTICATE [totp]
S: WELCOME

In this case, the router checks each result sequentially before sending out the next challenge. At any point it can abort the connection.

I am not sure where the server-last can be sent. It can either be a separate message, or it can be bundled with the next challenge. Both of these seem like poor solutions.

2. Send challenges out in parallel.
This is more general. When the client sends a HELLO, the server would send all challenges at the same time. So, the TOTP and the WAMP-SCRAM challenge would be issued simultaneously. They can be distinguished by the 'extra' parameter. The client can respond to the challenges in any order it wishes. On the router side, there would be a master authenticator method. It would get called after each authentication attempt. For example, the authentication flow would look like:

The server would have three methods:
WAMPAuthenticator.onAuthenticate()
TOTPAuthenticator.onAuthenticate()
MasterAuthenticator.onAuthenticationResult()

C: HELLO
S: CHALLENGE [wamp-scram]
S: CHALLENGE [totp]
S: CHALLENGE [ticket]
C: AUTHENTICATE [wamp-scram]

WAMPAuthenticator.onAuthenticate() is called. It returns (success, wamp-scram-server-proof).

S: COUNTER-AUTHENTICATE [wamp-scram] (This sends the server proof. Unfortunately, it is not possible to bundle this with anything and it requires another message type.)

Additionally, the result of onAuthenticate() is sent to the MasterAuthenticator.

MasterAuthenticator decides it is not ready to authenticate the user quite yet, and returns AuthenticationPending. The server waits.

C: AUTHENTICATE [totp]

TOTPAuthenticator.onAuthenticate is called. It returns success. This is sent to the MasterAuthenticator. It decides it has enough evidence to authenticate the user.

S: WELCOME

**Comment by oberstet** [comment]
More ideas:

```
C: HELLO(authmethods = [cookie, scram-totp, scram, cra])

S: CHALLENGE(AuthMethod = "scram-totp")

C: CHALLENGE(AuthMethod = "scram-totp", Extra = {"scram": client-first})

S: AUTHENTICATE(Signature = server-first)

C: AUTHENTICATE(Signature = client-final)

S: CHALLENGE(AuthMethod = "totp", Extra = {"scram": server-final})

C: AUTHENTICATE(Signature = otp-value)

S: WELCOME
```

**Comment by oberstet** [comment]
We should learn (or even use) from SASL http://en.wikipedia.org/wiki/Simple_Authentication_and_Security_Layer

**Comment by ysobolev** [comment]
SASL RFC: http://www.ietf.org/rfc/rfc4422.txt

**Comment by Paranaix** [comment]
Again only slightly related, but if we are already reworking the authentication process I would opt again for allowing the router to pass an *extra* field in the *WELCOME* message (e.g to give a client a session token). If the majority agrees (or atleast don't disagree :smile:), I will create a new issue.

**Comment by oberstet** [comment]
@Paranaix yes, pls file a new issue, ideally describing practical use case and semantics.

**Comment by ysobolev** [comment]
Here is what we have so far (from IRC):

Authentication via SCRAM first, then OTP
--------------------------------------------------------
C: HELLO (scram-totp)

S: CHALLENGE (scram)
C: AUTHENTICATE (client-first) 
S: CHALLENGE (server-first)
C: AUTHENTICATE (client-last) 
S: CHALLENGE (server-last)
C: AUTHENTICATE (empty)

S: CHALLENGE (otp) 
C: AUTHENTICATE (otp)

S: WELCOME

Authentication via OTP first, then SCRAM
--------------------------------------------------------
C: HELLO (totp-scram)

S: CHALLENGE (otp) 
C: AUTHENTICATE (otp)

S: CHALLENGE (scram)
C: AUTHENTICATE (client-first) 
S: CHALLENGE (server-first)
C: AUTHENTICATE (client-last) 
S: CHALLENGE (server-last)
C: AUTHENTICATE (empty)

S: WELCOME

The reason for the empty AUTHENTICATE message is a) it is in line with SASL and b) it allows every "challenge" to be answered. Part b) is very important as it allows authentication mechanisms to be atomic. Every one starts with a CHALLENGE and ends with an AUTHENTICATE.

Comments?

@Paranaix: I second your idea. It could be possible to slip it in among the challenge-response exchanges too. I currently acquire a session cookie via an RPC call after authentication and think it is a little ugly.

**Comment by oberstet** [comment]
@ysobolev I like it:

* it does not introduce new message types
* it does stick to a `CHALLENGE` is a request to be answered by `AUTHENTICATE`
* it does not piggyback authentication method specific stuff on `HELLO` or `WELCOME`
* message exchanges are running serially (*no* parallel CHALLENGE/AUTHENTICATE exchanges)
* it is backwards compatible with the existing schemes

And: it allows use to combine any set of atomic authentication methods (like `cookie`, `scram`, `totp`, ..) into ad-hoc multi-factor authentication methods without writing new code.

I think that last one (if we really can make it fly) is pretty cool.

Compared to PAM, the authentication methods run in strict order and the first failing one fails the complete chain.

**Comment by ysobolev** [comment]
PAM has other nice features such as 'sufficient', 'binding', 'requisite', and 'required' modules. For example, if a 'sufficient' module succeeds and no previous modules have failed, it returns immediately. This is, however, a router specific issue and can be implemented without breaking the protocol spec. I would just add that a WELCOME or ABORT can come at any time between authentication mechanisms.

Here is a use case:

HELLO (cookie-scram-totp)

The router understands that cookie is sufficient. If it succeeds, the user is immediately logged in and does not have to finish the authentication process. If it fails, authentication can continue because 'cookie' is only sufficient, and not required.

Anyway, as I said already, this is a router specific issue, and can also be solved by having the client attempt two different authentications.

**Comment by oberstet** [comment]
@ysobolev interestingly, this points to another aspect which we should better describe:

The client in `HELLO` sends a *list* of authentication methods it supports. The router chooses *one* from the list.

The cookie case above:

`HELLO(authmethods = ["cookie", "scram-totp"]`

means, the router already can login the client immediately if there is a suitable cookie.

I would have expected that

`HELLO(authmethods = ["cookie-scram-totp"])`

means **all 3** must succeed, and the whole chain is failed upon the *first failing*.

**Comment by ysobolev** [comment]
@oberstet  Hm... I see your point. I think you are right, `cookie-scram-totp` should be all 3. Right now the crossbar router loops over available auth methods and picks the first viable one. This means after trying `cookie` in your case, it would give up if, say, the cookie was expired. It would make sense to try the remaining methods.

**Comment by Paranaix** [comment]
@ysobolev 
Does it really make sense to automatically try all available authentication methods (if some fail)? IMO thats counter-intuitive, as a failed authentication is usually an event one wants to handle (e.g display an error-message). The only authmethod coming to my mind which can be silently ignored if its fails is cookie based auth. In any case, I think its the responsibility of the client to continue authentication in such a case. However then, we can also simply restart the whole authentication process (*HELLO*) but without the failed method.

**Comment by oberstet** [comment]
Yeah, the idea was to allow to express preferences, but have fallbacks.

@Paranaix Crossbar.io will start the first authentication method announced in the client provided list that is configured in the Crossbar.io node configuration. It will not try other methods if the former selected one fails.

Like a client requesting auth. methods:

1. ["cookie", "wampcra"]
2. ["cookie", "scram", "wampcra"]
3. ["cookie", "scram-totp", "scram", "wampcra"]
4. ["cookie-totp", "scram-totp"]

But it could of course also specify 3-factor authentication for the ultra-paranoid;)

5. ["cert-scram-totp"]

where a client will have a pre-authenticated transport from verified from a TLS client certificate, then do a SCRAM password authentication, and finally ask for a one-time password.

**Comment by ysobolev** [comment]
@Paranaix This is true. I take it back.

**Comment by oberstet** [comment]
Probably we should also add text describing the difference between **transport-level authentication** methods like

* "cookie",
* "cert" (from TLS client certificates),
* "os" (from OS user account with Unix domain sockets)

and **handshake-based authentication** methods which are implemented via actual WAMP message exchange like

* "wampcra"
* "ticket"
* "scram"
* "totp"

**Comment by ysobolev** [comment]
OK, so maybe both of these approaches is reasonable.

Option A: The client tries `[cookie]` and if that fails (producing a message "Expired cookie."), tries `[scram-totp, cert]` meaning it is willing to authenticate with either one and a failure on one is a complete failure.

Option B: The client tries `[cookie, scram-totp, cert]` meaning that it wants to authenticating with one of those and keep trying them in order until all fail (which produces an error message "All authentication methods failed.").

**Comment by oberstet** [comment]
@ysobolev Crossbar.io works slightly different: when clients announces ["cookie", "wampcra"] and the browser sends a cookie, Crossbar.io will first inspect the cookie. If is is valid (non-expired), the authentication information (authid, authrole) stored with the cookie in the local Crossbar.io node is set as "transport level auth information". Later, when the client sends HELLO with ["cookie", "wampcra"], the WAMP message level processing will see the session's transport is already authenticated and immediately WELCOME the client with the stored auth info. When there is no transport auth in place, the WAMP-CRA message exchange will kick in. On successful authentication, when no cookie was set (or expired), the cookie is updated for the authentication info.

**Comment by ysobolev** [comment]
Ah, neat. I did not realize there was true transport level authentication. I guess replace all instances of `cookie` with `plain` or something in my comment above (and lose the `cert`).

**Comment by oberstet** [comment]
We should also describe a commonly used client announced list of authentication methods:

["cookie", "wampcra", "anonymous"]

The reason to include an "anonymous" fallback is for a user client to perform some kind of signup or initial registration.

Of course a client that "authenticated" using method "anonymous" will get some kind of "anonymous" authentication role assigned, and by virtue of that, a limited set of permissions.

**Comment by oberstet** [comment]
Using multi-factor designators like `scram-totp` won't work, as we now allow to have suffix parts designate algo variants, like `scram-sha256`.

Maybe we really need to allow for lists of lists here:

```python
[['cookie', 'wampcra', 'ticket'], ['cryptosign', 'ticket']]
```

Client announcing that it is willing to authenticate via specific 3-FA or 2-FA combindations.

**Comment by ecorm** [comment]
> Using multi-factor designators like scram-totp won't work, as we now allow to have suffix parts designate algo variants, like scram-sha256.

Over in #135, @meejah and I are now in agreement that the SCRAM variant would be best specified as a separate `"algorithm"` attribute. If you think we should keep the suffix part, perhaps you can explain why in #135.
